services/agents/channel/channel.go

```diff
@@ -280,23 +299,46 @@ func (c *Channel) subscribe(id uint32) chan agentpb.AgentResponsePayload {
 	return ch
 }
 
-func (c *Channel) publish(id uint32, resp agentpb.AgentResponsePayload) {
+func (c *Channel) removeResponseChannel(id uint32) (chan Response, error) {
 	c.rw.Lock()
-	if c.responses == nil { // Channel is closed, no more publishing
+	if c.responses == nil { // Channel is closed
 		c.rw.Unlock()
-		return
+		return nil, nil
 	}
 
 	ch := c.responses[id]
+	c.rw.Unlock()
 	if ch == nil {
-		c.rw.Unlock()
-		c.close(errors.WithStack(fmt.Errorf("no subscriber for ID %d", id)))
-		return
+		return nil, errors.WithStack(fmt.Errorf("no subscriber for ID %d", id))
 	}
-
+	c.rw.Lock()
 	delete(c.responses, id)
 	c.rw.Unlock()
-	ch <- resp
+	return ch, nil
+}
+
+// cancel sends an error to the subscriber.
+func (c *Channel) cancel(id uint32, err error) {
+	if ch, _ := c.removeResponseChannel(id); ch != nil {
+		ch <- Response{Error: err}
+	}
+}
+
+func (c *Channel) publish(id uint32, status *protostatus.Status, resp agentpb.AgentResponsePayload) {
+	if status != nil {
+		l := logger.Get(c.s.Context())
+		l.Errorf("got response %v with status %v", resp, status)
+		c.cancel(id, grpcstatus.FromProto(status).Err())
+		return
+	}
+	ch, err := c.removeResponseChannel(id)
+	if err != nil {
+		c.close(err)
```

Comment:
Should we really close the connection?

Reply:
Closing the channel is the current behavior of the publish when there is no subscriber with given id. If we should ignore such publish calls, then we should not close it here. Do you prefer to leave it open it such case?

Reply:
Ok, I will implement it then.